src/Compilers/Core/Portable/InternalUtilities/OneOrMany.cs

```diff
@@ -105,6 +105,44 @@ public T Current
                 get { return _collection[_index]; }
             }
         }
+
+        public static bool operator ==(OneOrMany<T> first, OneOrMany<T> second)
+        {
+            return first.Equals(second);
+        }
+
+        public static bool operator !=(OneOrMany<T> first, OneOrMany<T> second)
+        {
+            return !(first == second);
+        }
+
+        public bool Equals(OneOrMany<T> other)
+        {
+            if (Count != other.Count)
+            {
+                return false;
+            }
+
+            for (int i = 0; i < Count; i++)
+            {
+                if (this[i].Equals(other[i]))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        public override bool Equals(object obj)
+        {
+            return obj is OneOrMany<T> && Equals((OneOrMany<T>)obj);
+        }
+
+        public override int GetHashCode()
+        {
+            return base.GetHashCode();
```

Comment:
`Add` needs to be changed to use `_one` rather than `_many` when adding to an empty set. Otherwise, `base.GetHashCode()` will return different values for the `_one` and `_many.Length == 1` cases, even though those will compare `Equals`. We may be better off not allowing an empty `OneOrMany<T>`.